[KEP-5075] DRAConsumableCapacity documentation (#51479)
kubernetes/website: content/en/docs/concepts/scheduling-eviction/dynamic-resource-allocation.md

[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
is enabled in the kube-apiserver and kube-scheduler.

## Consumable capacity

{{< feature-state feature_gate_name="DRAConsumableCapacity" >}}

> Device driver can set `allowMultipleAllocations` field added in `.spec.devices` of `ResourceSlice` to allow allocating that device to multiple independent ResourceClaims or to multiple requests within a ResourceClaim.

Reviewer: ```suggestion
The consumable capacity feature allows the same devices to be consumed by multiple independent ResourceClaims, with the Kubernetes scheduler
managing how much of the device's capacity is used up by each claim. This is analogous to how Pods can share
the resources on a Node; ResourceClaims can share the resources on a Device.

The device driver can set `allowMultipleAllocations` field added in `.spec.devices` of `ResourceSlice` to allow allocating that device to multiple independent ResourceClaims or to multiple requests within a ResourceClaim.
```

Users can set `capacity` field added in `spec.devices.requests` of `ResourceClaim` to specify the device resource requirements for each allocation.

> For the device that allows multiple allocations, the requested capacity is drawn from—or consumed from—its total capacity, a concept known as **consumable capacity**.

Reviewer: ```suggestion
For the device that allows multiple allocations, the requested capacity is drawn from — or consumed from — its total capacity, a concept known as **consumable capacity**.
```

> Then, the scheduler ensures that the aggregate consumed capacity across all claims does not exceed the device’s overall capacity and complies with any configured `requestPolicy` constraints in `.spec.devices[*].capacity`, if specified.

Reviewer: ```suggestion
Then, the scheduler ensures that the aggregate consumed capacity across all claims does not exceed the device’s overall capacity. Furthermore, driver authors can use the `requestPolicy` constraints on individual device capacities to control how those capacities are consumed. For example, the driver author can specify that a given capacity is only consumed in increments of 1Gi.
```

Here is an example of a network device which allows multiple allocations and contains
a consumable bandwidth capacity.

```yaml
kind: ResourceSlice

> apiVersion: resource.k8s.io/v1beta2

Reviewer: ```suggestion
apiVersion: resource.k8s.io/v1
```

metadata:
  name: resourceslice
spec:
  nodeName: worker-1
  pool:
    name: pool
    generation: 1
    resourceSliceCount: 1
  driver: dra.example.com
  devices:
  - name: eth1
    allowMultipleAllocations: true
    attributes:
      name:
        string: "eth1"
    capacity:
      bandwidth:
        requestPolicy:
          default: "1M"
          validRange:
            min: "1M"
            step: "8"
        value: "10G"
```

The property of `allowMultipleAllocations` can be used in CEL
and the consumable capacity can be requested as shown in the below example.

```yaml
apiVersion: resource.k8s.io/v1beta2
kind: ResourceClaimTemplate
metadata:
  name: bandwidth-claim-template
spec:
  spec:
    devices:
      requests:
      - name: req-0
        exactly:
        - name:
          deviceClassName: resource.example.com
          selectors:
          - cel:
              expression: |-

>                 device.allowMultipleAllocations == true

Reviewer: This is not *required*, right? This would limit the chose of device to ONLY those with consumable capacity. Let's leave it out of the basic example. Then, after your example allocation, I would add the paragraph:

In this example, a multiply-allocatable device was chosen. However, *any* `resource.example.com` device with at least the requested 1G bandwidth could have met the requirement. If a non-multiply-allocatable device were chosen, the allocation would have resulted in the entire device. To force the use of a *only* multiply-allocatable devices, you can use the CEL criteria `device.allowMultipleAllocations == true`.
